pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageContainerImpl.java

```diff
@@ -108,6 +108,8 @@ public boolean add(MessageImpl<?> msg, SendCallback callback) {
                 }
             } catch (Throwable e) {
                 log.error("construct first message failed, exception is ", e);
+                producer.semaphoreRelease(getNumMessagesInBatch());
```

Comment:
Maybe we need to throw this exception and then let the producer handle this logic.